### PR TITLE
build: update build process to conform to Angular Package Format

### LINKED ISSUE #49: refactor: change project structure to Angular Library format
Currently the package is built in such a way that a lot of files are duplicated as both ts and js files. The project should be refactored to use a build process similar to Angular core, Material, and Flex-Layout, with multiple bundles for ES5 and ES2015.

At the very least, it would allow for top-level imports instead of `import * from 'preboot/src'` as currently required.

### LINKED ISSUE #52: ng serve & ng e2e fails when using preboot with angular CLI
* **I'm submitting a ...**
- [X] bug report
- [ ] feature request

* **Which parts of preboot are affected by this issue?**
- [ ] server side
- [ ] client side
- [ ] inline
- [X] build process
- [ ] docs
- [ ] tests

* **Do you want to request a *feature* or report a *bug*?**
Bug

* **What is the current behavior?**
`ng serve` and `ng e2e`  are failing with the following error log in the console:

```
C:\Users\PrebootPC\Desktop\Projects\Preboot\Preboot>ng e2e
** NG Live Development Server is listening on localhost:49152, open your browser on http://localhost:49152/ **
Date: 2017-11-06T10:04:48.102Z
Hash: 3d6f41a2627365d2252f
Time: 11686ms
chunk {auth.module} test.module.chunk.js, auth.module.chunk.js.map () 148 kB  [rendered]
chunk {email.module} auth.module.chunk.js, email.module.chunk.js.map () 3.61 kB  [rendered]
chunk {inline} inline.bundle.js, inline.bundle.js.map (inline) 5.83 kB [entry] [rendered]
chunk {main} main.bundle.js, main.bundle.js.map (main) 82.5 kB [initial] [rendered]
chunk {polyfills} polyfills.bundle.js, polyfills.bundle.js.map (polyfills) 217 kB [initial] [rendered]
chunk {settings.module} settings.module.chunk.js, settings.module.chunk.js.map () 6.87 kB  [rendered]
chunk {storage.module} welcome.module.chunk.js, welcome2.module.chunk.js.map () 3.66 kB  [rendered]
chunk {styles} styles.bundle.js, styles.bundle.js.map (styles) 77.1 kB [initial] [rendered]
chunk {tasks.module} home.module.chunk.js, core.module.chunk.js.map () 51.7 kB  [rendered]
chunk {updates.module} test2.module.chunk.js, updates.module.chunk.js.map () 3.66 kB  [rendered]
chunk {vendor} vendor.bundle.js, vendor.bundle.js.map (vendor) 7.39 MB [initial] [rendered]

ERROR in ./node_modules/preboot/src/browser/browser-preboot.module.ts
Module build failed: Error: C:\Users\PrebootPC\Desktop\Projects\Preboot\Preboot\node_modules\preboot\src\browser\browser-preboot.module.ts is not part of the compilation output. Please check the other error messages for details.
    at AngularCompilerPlugin.getCompiledFile (C:\Users\PrebootPC\Desktop\Projects\Preboot\Preboot\node_modules\@ngtools\webpack\src\angular_compiler_plugin.js:629:23)
    at plugin.done.then (C:\Users\PrebootPC\Desktop\Projects\Raya\raya\node_modules\@ngtools\webpack\src\loader.js:467:39)
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:188:7)
 @ ./src/app/app.module.ts 15:0-82
 @ ./src/main.ts
 @ multi webpack-dev-server/client?http://0.0.0.0:0 ./src/main.ts

webpack: Failed to compile.
Build did not succeed. Please fix errors before running e2e task
```

* **If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem** by creating a github repo.

* **What is the expected behavior?**
app should serve with `ng serve`
e2e tests should pass.

* **Please tell us about your environment:**

- Browser: [Chrome 63]
- Language: [TypeScript 2.6.1]
- OS:  [Windows]
- Platform: [NodeJs]

image

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. stackoverflow, gitter, etc)

`ng build --prod` & `ng test` do not emit the error described.

### PR DESCRIPTION
**Pull request by CaerusKaru**
This is the improved version of #50. Preboot will now use the same bundling method that Angular Material and Flex-Layout use.

* update build process to use rollup and Angular compiler (`ngc`) to produce ES5 and ES2015 bundles
* remove the inline script builder in favor of run-time compilation
* combine `ServerPrebootModule` and `BrowserPrebootModule` into a singular `PrebootModule`, also combine `PrebootReplayOptions` and `PrebootRecordOptions` into `PrebootOptions`
* Remove relative import in favor of root import, `import {...} from 'preboot/server';` --> `import {...} from 'preboot';`
* Deprecate the minification feature due to removal of inline script builder
* Make nonce import cleaner
* Update preboot deps to Angular v5

BREAKING CHANGE:

* Any imports of `ServerPrebootModule` or `BrowserPrebootModule` need to be removed in favor of `PrebootModule`, and all relative imports need to be changed to import directly from `preboot`
* The minification option has been removed. Official guidance is for end-users to perform minification manually on the rendered HTML

Fixes #49 
Fixes #52

### CODE REVIEW

**Comment by jeffwhelpley** [comment]
Btw, if you want to jump on a hangout to work on this together at any point, just let me know.

**Comment by CaerusKaru** [comment]
You're having issues with building because I suspect your version of rollup is too high. They don't follow semver, so the latest version (0.51.x) is incompatible with build config for the version that's used here and by the other Angular libraries (0.41.x). I'll update the package.json to reflect this. As for the tsc issues, my build runs fine, what version of TypeScript are you using? Again, since Angular 5 compiler still only supports TypeScript 2.4.2, that's the version I'm using here (I'll change this in the package.json too).

I'd be happy to hop on a Hangout, the only issue I'm still facing is the e2e testing bug. Everything builds successfully, but the EventReplayer won't kick in on its own during bootstrap (although it works fine in real applications).

File: integration/package.json [hunk 1]
```diff
@@ -0,0 +1,70 @@
+{
```
[the remark this answers is not in the document]
Comment (CaerusKaru) [reply integration/package.json]: Added, and removed package-lock.json

**Comment by CaerusKaru** [comment]
@jeffwhelpley Just for the record, here's my build process at the moment (I'll cut it down for the official release):

1. In `src`, run `npm i` and `npm run prerelease`
2. In `integration`, run `npm run build` and `npm run e2e:aot`

You should see at the point that the e2e passes the first two tests, but fails on the next two

**Comment by CaerusKaru** [comment]
@jeffwhelpley This is ready to be merged, I resolved the final e2e issue (which turned out to be a good thing because the timing on the EventReplayer was very off).

The official commands to run the build are close to the old version:

`npm run prerelease && npm run e2e`

This will build the library and run the e2e tests. At some point we'll probably want to incorporate some unit tests as well, but that's been left as a stub for now.

Let me know what you think, and if any other changes are needed.

**Comment by CaerusKaru** [comment]
@jeffwhelpley Just touching up the spec type issues you noted above (I finally read through and realized my build wasn't catching those type errors for some reason). I'll ping you when this is resolved, which should be in the next few minutes (sorry for the ping spam).

**Comment by CaerusKaru** [comment]
@jeffwhelpley type issues resolved, everything should be set

**Comment by jeffwhelpley** [comment]
@CaerusKaru ok, this is definitely going in the right direction but I still have a few more questions. Rather than go back an forth on here, let's just do a Google Hangout Monday or Tuesday night this week to wrap things up. Send me your email to jeff at gethuman dot com.

**Comment by mgol** [comment]
Is there anything else blocking this PR?

**Comment by jeffwhelpley** [comment]
@mgol apologies, we are still working out one issue. We will get this resolved today.

**Comment by jeffwhelpley** [comment]
OK, @mgol @CaerusKaru I merged in this PR and published it under `6.0.0-beta.0`. Please try it out and let me know if you run into any issues.

**Comment by misticwonder** [comment]
There is an issue with styles. After switching to front part. They have no styles applied. I.e. 
```
<app-root style="display: none;"><app-home-info _ngcontent-c1=""><app-root>
<app-root><app-home-info _ngcontent-c2=""><app-root>
```
But styles only for _ngcontent-c1

**Comment by CaerusKaru** [comment]
@misticwonder Can you link to a minimal reproduction repo so we can check this out?

**Comment by misticwonder** [comment]
I've found a problem. In my project appRef.isStable is never true.

**Comment by misticwonder** [comment]
There is another issue:
When injector.get(ApplicationInitStatus).donePromise resolves - angular removes all styles with 'ng-transition' from the header. But appRef.isStable becomes true later in my app. This leads to the messing on the page for time between this 2 events.

**Comment by CaerusKaru** [comment]
@misticwonder Please open a separate issue with a minimal reproduction instead of appending to this PR

**Pull request merged**